pkg/util/intern/string.go

```diff
@@ -0,0 +1,136 @@
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// code adapted from go4.org/intern
+
+// Package intern lets you make smaller comparable values by boxing
+// a larger comparable value (such as a 16 byte string header) down
+// into a globally unique 8 byte pointer.
+//
+// The globally unique pointers are garbage collected with weak
+// references and finalizers. This package hides that.
+package intern
+
+import (
+	"runtime"
+	"sync"
+	"unsafe"
+)
+
+// A StringValue pointer is the handle to the underlying string value.
+// See Get how Value pointers may be used.
+type StringValue struct {
+	_           [0]func() // prevent people from accidentally using value type as comparable
+	cmpVal      string
+	resurrected bool
+}
+
+// Get the underlying string value
+func (v *StringValue) Get() string {
+	return v.cmpVal
+}
+
+// StringInterner interns strings while allowing them to be cleaned up by the GC.
+// It can handle both string and []byte types without allocation.
+type StringInterner struct {
+	mu     sync.Mutex
+	valMap map[string]uintptr
+}
+
+// NewString creates a new StringInterner
+func NewString() *StringInterner {
+	return &StringInterner{
+		valMap: make(map[string]uintptr),
+	}
+}
+
+//go:nocheckptr
+// GetString returns a pointer representing the string k
+//
+// The returned pointer will be the same for GetString(v) and GetString(v2)
+// if and only if v == v2. The returned pointer will also be the same
+// for a byte slice with same contents as the string.
+func (s *StringInterner) GetString(k string) *StringValue {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var v *StringValue
+	if addr, ok := s.valMap[k]; ok {
+		//goland:noinspection GoVetUnsafePointer
+		v = (*StringValue)((unsafe.Pointer)(addr))
+		v.resurrected = true
+		return v
+	}
+
+	v = &StringValue{cmpVal: k}
+	runtime.SetFinalizer(v, s.finalize)
+	s.valMap[k] = uintptr(unsafe.Pointer(v))
+	return v
+}
+
+//go:nocheckptr
+// Get returns a pointer representing the []byte k
+//
+// The returned pointer will be the same for Get(v) and Get(v2)
+// if and only if v == v2. The returned pointer will also be the same
+// for a string with same contents as the byte slice.
+func (s *StringInterner) Get(k []byte) *StringValue {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var v *StringValue
+	// the compiler will optimize the following map lookup to not alloc a string
+	if addr, ok := s.valMap[string(k)]; ok {
+		//goland:noinspection GoVetUnsafePointer
+		v = (*StringValue)((unsafe.Pointer)(addr))
+		v.resurrected = true
+		return v
+	}
+
+	v = &StringValue{cmpVal: string(k)}
+	runtime.SetFinalizer(v, s.finalize)
+	s.valMap[string(k)] = uintptr(unsafe.Pointer(v))
+	return v
+}
+
+func (s *StringInterner) finalize(v *StringValue) {
+	s.mu.Lock()
```

Comment:
Since all finalizers run on a single goroutine, is there a risk a highly loaded interner would slow down or prevent other finalizers from making progress?

Reply:
The docs make it sound like each finalizer gets its own goroutine
> When the garbage collector finds an unreachable block with an associated finalizer, it clears the association and runs finalizer(obj) in a separate goroutine.